nidx/nidx_paragraph/src/query_parser/fuzzy_parser.rs

```diff
@@ -0,0 +1,141 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+//
+
+use tantivy::Term;
+use tantivy::query::AllQuery;
+use tantivy::query::BooleanQuery;
+use tantivy::query::Occur;
+use tantivy::query::PhraseQuery;
+use tantivy::query::Query;
+use tantivy::query::TermQuery;
+use tantivy::schema::IndexRecordOption;
+
+use crate::fuzzy_query::FuzzyTermQuery;
+use crate::schema::ParagraphSchema;
+use crate::search_query::SharedTermC;
+
+use super::tokenizer::Token;
+
+/// Minimum length required to be considered a fuzzy word. Words with smaller
+/// length than this are considered too short to be fuzzy. This is done to avoid
+/// too much noise from short terms.
+const MIN_FUZZY_LEN: usize = 3;
+
+/// Minimum length required to be considered a fuzzy prefix. This is again don
+/// eto avoid too much noise from short term.
+const MIN_FUZZY_PREFIX_LEN: usize = 4;
+
+/// Levenshtein distance used by all fuzzy terms.
+pub const FUZZY_DISTANCE: u8 = 1;
+
+/// Convert a tokenized query into a tantivy fuzzy query
+///
+/// A fuzzy query will match similarly to a keyword query but some terms will be
+/// searched with certain Levenshtein distance. Quoted and excluded terms will
+/// remain the same, but literals will be elegible to be fuzzy. All long enough
+/// literals will be searched as fuzzy. The last literal can be searched as a
+/// fuzzy prefix if a suggest-like feature is desired.
+///
+pub fn parse_fuzzy_query(query: &[Token], term_collector: SharedTermC, last_literal_as_prefix: bool) -> Box<dyn Query> {
+    // We use usize::MAX as a discriminant when we don't want/care about the
+    // last index. This value won't match as there's no query that long.
+    let last_literal_index = if last_literal_as_prefix {
+        query.iter().enumerate().fold(
+            usize::MAX,
+            |acc, (idx, token)| {
+                if matches!(token, &Token::Literal(_)) { idx } else { acc }
+            },
+        )
+    } else {
+        usize::MAX
+    };
+
+    let mut subqueries = vec![];
+    let schema = ParagraphSchema::new();
+
+    for (i, item) in query.iter().enumerate() {
+        match item {
+            Token::Literal(literal) => {
+                let term = Term::from_field_text(schema.text, literal);
+                let distance = FUZZY_DISTANCE;
+                let transposition_cost_one = true;
+
+                let q: Box<dyn Query>;
+                if literal.len() < MIN_FUZZY_LEN {
+                    // to avoid noise, we don't want to match too short terms as fuzzy
+                    q = Box::new(TermQuery::new(term, IndexRecordOption::Basic));
+                } else if last_literal_as_prefix && (i == last_literal_index) && literal.len() >= MIN_FUZZY_PREFIX_LEN {
```

Comment:
`literal.len() >= MIN_FUZZY_PREFIX_LEN` is guaranteed to always be `false` here (otherwise, it should have gone into the first branch of the if). You can remove that part of the condition.

Reply:
Ah no, `MIN_FUZZY_LEN` and `MIN_FUZZY_PREFIX_LEN` are different things.

Reply:
We could even decide to use the same constant, I don't think the exact number matters too much anyway :stuck_out_tongue: